fixed_bed_model.py

```diff
@@ -0,0 +1,2630 @@
+    elif m.diff==-1:
+        if dv=='k':
+            return (m.FCO2['base',p[-1], t] - m.FCO2['backward_k',p[-1],t])/m.eps/m.para
+        elif dv=='ua':
+            return (m.FCO2['base',p[-1],t] - m.FCO2['backward_ua',p[-1],t])/m.eps/m.ua_init
+    elif m.diff==2:
+        if dv=='k':
+            return (m.FCO2['forward_k', p[-1],t] - m.FCO2['backward_k',p[-1],t])/m.eps/m.para/2
+        elif dv=='ua':
+            return (m.FCO2['forward_ua',p[-1],t] - m.FCO2['backward_ua',p[-1],t])/m.eps/m.ua_init/2
+
+def jac_Tend_ele(m,dv,t):
+    '''
+    Calculate the Jacobian elements 
+    Arguments:
+        dv: the design variables, k for k_trans, ua for UA
+        t: timepoints 
+    Return: 
+        jac['k',t] is the Jacobian of k_trans, jac['ua', t] is the Jacobian of UA
+    '''
+    p = []
+    for posi in m.zgrid:
+        p.append(posi)
+    
+    if m.diff ==1:
+        if dv=='k':
+            return (m.temp['forward_k', p[-1], t] - m.temp['base',p[-1],t])/m.eps/m.para
+        elif dv=='ua':
+            return (m.temp['forward_ua',p[-1],t] - m.temp['base',p[-1],t])/m.eps/m.ua_init
+    elif m.diff==-1:
+        if dv=='k':
+            return (m.temp['base',p[-1], t] - m.temp['backward_k',p[-1],t])/m.eps/m.para
+        elif dv=='ua':
+            return (m.temp['base',p[-1],t] - m.temp['backward_ua',p[-1],t])/m.eps/m.ua_init
+    elif m.diff==2:
+        if dv=='k':
+            return (m.temp['forward_k', p[-1],t] - m.temp['backward_k',p[-1],t])/m.eps/m.para/2
+        elif dv=='ua':
+            return (m.temp['forward_ua',p[-1],t] - m.temp['backward_ua',p[-1],t])/m.eps/m.ua_init/2
+
+        
+def feed_rule(m):
+    '''The inlet mole fraction should add up to 1. Applied for continuous optimization problem 
+    '''
+    return sum(m.yfeed[i] for i in m.COMPS) == 1
+
+    
+def ObjRule_opt_k(m):
+    '''objective function for the continuous DoE problem
+    '''
+    objfunc = -sum(m.jac['k', t]**2 for t in m.t)
+    return objfunc    
+
+def ObjRule_A(m):
+    '''
+    A-optimality objective function
+    '''
+    fim00 = sum(m.jac['k',t]**2 for t in m.t) + sum(m.jac_tend['k',t]**2 for t in m.t)
+    #fim01 = sum(m.jac['k',p[-1],t]*m.jac['ua',t] for t in m.t)
+    fim11 = sum(m.jac['ua',t]**2 for t in m.t) +sum(m.jac_tend['ua',t]**2 for t in m.t)
+    # use -trace for minimize, to maximize trace
+    trace = -fim00 - fim11
+    return trace
+     
+def ObjRule_D(m):
+    '''
+    D-optimality objective function
+    '''
+    fim00 = sum(m.jac['k',t]**2 for t in m.t)+ sum(m.jac_tend['k',t]**2 for t in m.t)
+    fim01 = sum(m.jac['k',t]*m.jac['ua',t] for t in m.t) + sum(m.jac_tend['k',t]*m.jac_tend['ua',t] for t in m.t)
+    fim11 = sum(m.jac['ua',t]**2 for t in m.t)+sum(m.jac_tend['ua',t]**2 for t in m.t)
+    # use -det for minimize, to maximize det
+    det = fim01**2 - fim00*fim11
+    return det
+    
+def ObjRule_con(m):
+    '''objective function for the square problem
```

Comment:
obsolete? Or is this used for initialization? If so, add a comment explaining this function is used for initialization.